### PR TITLE
[mqtt] Ring buffer write succeeds when no space to fit a new insert

### PR DESCRIPTION
**Pull request by dmolokanov**
There was an issue when ring buffer calculated free space available to insert a new publication.

### CODE REVIEW

**Review by nyanzebra — APPROVED** [review]

File: mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs [hunk 1]
```diff
@@ -200,15 +199,24 @@ impl StreamWakeableState for RingBuffer {
...
+        if write_index < read_index {
+            free_space %= self.max_file_size;
+        }
+        if free_space < total_size {
+            return Err(PersistError::RingBuffer(
+                RingBufferError::InsufficientSpace {
+                    free: free_space,
+                    required: total_size,
+                },
+            ));
```
Comment (nyanzebra) [inline mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs:219]: Just to make clear the issue for anyone else:
I thought that std File set_len was a max bound, but it is not.. from when I was using the MMAP crate it did work this way and an err would happen. What this change does is it makes sure that if we have both pointers at the same spot and no wrap around we calculate the remaining size correctly.

E.G. buffer is 200 and both pointers are 100 -> 100 + 200 - 100 => 200 now where before the result would be 0.

File: mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/error.rs [hunk 2]
```diff
@@ -29,6 +29,9 @@ pub enum RingBufferError {
     #[error("Buffer is full and messages must be drained to continue")]
     Full,
 
+    #[error("But has insufficient space to insert data: required: {required}b, but only {free}b available")]
```
Comment (vadim-kovalyov) [inline mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/error.rs:32]: typo? 0_o

Comment (vadim-kovalyov) [reply mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/error.rs:32]: What's the difference b/w `Full` and `InsufficientSpace`? It seems like we can have just one `InsufficientSpace` that covers everything.

Comment (nyanzebra) [reply mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/error.rs:32]: That is a fair point, I would prefer the InsufficientSpace over Full :)

Comment (dmolokanov) [reply mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/error.rs:32]: there is a special flag for 'Full' which made me add additional error variant.

Comment (nyanzebra) [reply mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/error.rs:29]: Yeah, the `Full` (and the current error in pipeline logs) was also useful for ignoring if we are full. But we should be able to switch to the `InsufficientSpace` version everywhere and get much more useful messages in logs

**Review by vadim-kovalyov — COMMENTED** [review]
Let me know when you fix the default feature :)

**Review by nyanzebra — APPROVED** [review]

File: mqtt/mqtt-bridge/src/messages.rs [hunk 3]
```diff
@@ -165,9 +165,17 @@ where
...
+                        Ok(_) => Ok(Handled::Fully),
+                        Err(
+                            err
+                            @
+                            PersistError::RingBuffer(RingBufferError::InsufficientSpace {
+                                ..
+                            }),
+                        ) => {
+                            error!(error = %err, "dropping incoming publication");
+                            Ok(Handled::Fully)
```
Comment (varunpuranik) [inline mqtt/mqtt-bridge/src/messages.rs:177]: Hmm, interesting that we return Ok here.

**Review by varunpuranik — APPROVED** [review]

File: mqtt/mqtt-bridge/src/messages.rs (hunk 3 above)
Comment (nyanzebra) [reply mqtt/mqtt-bridge/src/messages.rs:177]: There was a comment before, this seems to now be an error log. I think also based on current conversations we might want to add a TODO here for how we want to handle this in the future.

**Pull request merged**